### PR TITLE
[JIT] Refactor attributes to support buffers and parameters as first class citizens, add support for iterating over `named_buffers()`

### LINKED ISSUE #36211: [JIT] support self.named_buffers and self.named_parameters in TorchScript
## 🚀 Feature
 support self.named_buffers and self.named_parameters in TorchScript

## Motivation

Since `self.named_models` had been supported in torchscript [#29495 ]( https://github.com/pytorch/pytorch/pull/29495), it will be more convenient to support self.named_buffers and self.named_parameters. Projects like object detection and model compression use buffers to stores important things (e.g. anchor box and pruning mask).

## Pitch
```
  class MyMod(torch.nn.Module):
      def __init__(self):
          super(MyMod, self).__init__()
          self.register_buffer('x', torch.zeros(1))

      def forward(self, x):
          for name, buffer in self.named_buffers()
              x = x + buffer
          return x

  model = MyMod()
  torch.jit.script(model)
```

### PR DESCRIPTION
**Pull request by voznesenskym**
First part of #36211 - still a WIP, but asking for commentary to ensure this is the direction we want to go in.

### CODE REVIEW

**Review by suo — COMMENTED** [review]
Some high level thoughts, haven't had a chance to look deeply:
1. We should probably first tackle the change where we add buffers as a first-class concept on ClassType, separately from adding a `named_buffers()` iterator.
2. `is_parameter` and `was_registered_as_buffer` are mutually exclusive. Conceivably there should be an enum that expresses the three types of tensor of attributes directly (parameters, buffers, regular attributes).

**Comment by voznesenskym** [comment]
@suo - Good suggestion. Making it a first class citizen + adding an enum would less us have a single collection to manage what we have in each slot on ClassType. I'll go back to the drawing board and look at how we can cleanly manage all 3. Looking at it, perhaps we can remove attributeNames_,  attributeTypes_ parameterSlots_, and instead have a single `std::vector<AttributeType>` where `AttributeType` contains:

- the value for the enum you mentioned 
- the attribute name (string) (if relevant)
- the attributeType (TypePtr)

Thoughts?

**Comment by suo** [comment]
Yeah I think that's a good overall direction. Some minor bits of context that may help guide the data structure decision:
1. Runtime objects (instances of a ClassType) do not know about names, they store their attributes in a flat vector of IValue slots, to avoid hash table lookups in getattr. So we need a way of efficiently retrieving the slot of something by name on ClassType.
2. Objects may want to look up name/type/parameter status by slot.
3. It's okay to trade efficiency of retrieval in ClassType (which is a compile-time construct) for efficiency of retrieval in `ivalue::Object` (which is run-time). But that might not be relevant for this.

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -1600,6 +1600,39 @@ CAFFE2_API TypePtr tryEvalTypeVariables(TypePtr type, TypeEnv& type_env);
...
+
+  ATTRIBUTE_KIND_ENUM getKind() const {
+    return kind_;
+  }
+
+  TypePtr getType() const {
+    return attributeType_;
+  }
+
+  std::string getName() const {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1626]: Probably shouldn't be by copy.

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -1847,12 +1873,23 @@ struct CAFFE2_API ClassType : public NamedType {
...
   }
+
   bool is_parameter(size_t slot) const {
     TORCH_INTERNAL_ASSERT(
         is_module(), "asking for parameterSlots of non-Module");
-    return parameterSlots_->at(slot);
+    return attributes_.at(slot).getKind() == ATTRIBUTE_KIND_ENUM::PARAMETER;
+  }
+
+  bool is_buffer_written_attribute(size_t slot) const {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1889]: I'm seeing a few different ways of wording "is this a buffer?". Here it's `is_buffer_written_attribute`, and in `addAttribute` it's `wasRegisteredAsBuffer`. I think similar to parameters, it should just be `is_buffer` everywhere.

File: aten/src/ATen/core/type.cpp [hunk 3]
```diff
@@ -1166,19 +1182,21 @@ size_t ClassType::addAttribute(
         "Expecting parameter to have either None, Tensor or Optional[Tensor] type, but got: ",
         toString(type));
   }
-  if (is_module()) {
-    parameterSlots_->push_back(is_parameter);
+  if (was_registered_as_buffer) {
+      TORCH_INTERNAL_ASSERT(is_module(), "attempting to register a buffer on a non module");
+    TORCH_CHECK((type->kind() == TensorType::Kind),
```
Comment (suo) [inline aten/src/ATen/core/type.cpp:1187]: This check should be the same as the above. The PT API allows users to register `None`, `Tensor`, or `Optional[Tensor]` as a buffer.

File: aten/src/ATen/core/type.cpp [hunk 4]
```diff
@@ -1126,35 +1124,53 @@ void ClassType::checkNotExist(const std::string& name, const std::string& what)
...
+  what += (was_registered_as_buffer? "buffer" : "not buffer");
+  checkNotExist(name, what);
   if (!allow_any) {
-    checkNoAny(*this, what, name, type);
+    checkNoAny(*this, what.c_str(), name, type);
+  }
+
+  size_t slot = attributes_.size();
+
+  ATTRIBUTE_KIND_ENUM kind = ATTRIBUTE_KIND_ENUM::REGULAR_ATTRIBUTE;
```
Comment (suo) [inline aten/src/ATen/core/type.cpp:1163]: Should we consider just making `ATTRIBUTE_KIND_ENUM` something that the user sets? Hiding allows some weird behavior. For example: right now you can call `addAttribute` and have something be both a parameter and buffer—this code would silently mark the attribute as a parameter.

File: aten/src/ATen/core/jit_type.h [hunk 5]
```diff
@@ -1600,6 +1600,39 @@ CAFFE2_API TypePtr tryEvalTypeVariables(TypePtr type, TypeEnv& type_env);
 
 CAFFE2_API bool elementTypeCanBeInferredFromMembers(const TypePtr& elem_type);
 
+enum ATTRIBUTE_KIND_ENUM {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1603]: nit: while we are pretty loose with style, theoretically we follow fb style. So this should be `AttributeKindEnum`. 

Also: should use `enum class`

File: test/test_jit.py [hunk 6]
```diff
@@ -17139,6 +17139,46 @@ def forward(self):
         f = Foo()
         torch.jit.script(f)
 
+
+    def named_buffers_are_iterable(self):
```
Comment (suo) [inline test/test_jit.py:17143]: I think `unittest` expects test methods to start with `test_`?

File: test/test_jit.py [hunk 7]
```diff
@@ -17139,6 +17139,46 @@ def forward(self):
...
+                    names.append(name)
+                    vals.append(buffer + 2)
+                return names, vals
+
+            def forward(self, x):
+                return x
+
+        model = MyMod()
+        x = torch.jit.script(model)
+        z = self.getExportImportCopy(x)
```
Comment (suo) [inline test/test_jit.py:17171]: Does this work today? I would expect that we lose buffer status between save and load as we are not serializing it anywhere.

File: torch/csrc/jit/python/python_sugared_value.h [hunk 8]
```diff
@@ -176,6 +176,10 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
       const SourceRange& loc,
       Function& m);
 
+  std::shared_ptr<SugaredModuleDict> getSugaredNamedBufferDict(
```
Comment (suo) [inline torch/csrc/jit/python/python_sugared_value.h:179]: Probably we should rename `SugaredModuleDict` something more general, to reflact that it is not just being used as for `named_modules()` anymore.

File: aten/src/ATen/core/jit_type.h [hunk 9]
```diff
@@ -1847,12 +1873,23 @@ struct CAFFE2_API ClassType : public NamedType {
   }
 
   bool is_module() const override {
-    return bool(parameterSlots_);
+    return isModule_;
+  }
+
+  const std::vector<AttributeKind>& getAllAttributes() const {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1879]: nit: maybe `getAttributes()`?

File: aten/src/ATen/core/jit_type.h [hunk 10]
```diff
@@ -1896,27 +1933,31 @@ struct CAFFE2_API ClassType : public NamedType {
     return n.qualifiedName();
   }
 
+  void provideNewAttributeKind(AttributeKind attributeKind);
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1936]: nit: I think this name doesn't make it immediately clear that this function is basically the intenral impl of `addAttribute`. When I read "provide" I'm not really sure how it relates to the other stuff in this class, if that makes sense.

File: aten/src/ATen/core/jit_type.h [hunk 11]
```diff
@@ -1600,6 +1600,39 @@ CAFFE2_API TypePtr tryEvalTypeVariables(TypePtr type, TypeEnv& type_env);
 
 CAFFE2_API bool elementTypeCanBeInferredFromMembers(const TypePtr& elem_type);
 
+enum ATTRIBUTE_KIND_ENUM {
+  BUFFER,
+  PARAMETER,
+  REGULAR_ATTRIBUTE
+};
+
+struct CAFFE2_API AttributeKind {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1609]: nit: This represents an attribute, not a kind of attribute. I would say `AttributeKindEnum` should be `AttributeKind`, and `AttributeKind` should be `Attribute` (or `ClassAttribute` if `Attribute` is reserved elsewhere).

**Review by suo — CHANGES_REQUESTED** [review]
Nice! This is definitely on the right track. I left a somewhat unusual number of bikeshedding comments—this is a core API so we want to get the naming/concepts consistent.

File: aten/src/ATen/core/jit_type.h (hunk 2 above)
Comment (voznesenskym) [reply aten/src/ATen/core/jit_type.h:1889]: One is a method, so it asks the question, and another is a param, so it tells the function. is_buffer works for both though!

File: aten/src/ATen/core/type.cpp (hunk 4 above)
Comment (voznesenskym) [reply aten/src/ATen/core/type.cpp:1163]: I'd rather keep it as a detail of our system, and then provide safeguards. Users already signal if something is a buffer or parameter by how they register it with our module.

File: torch/csrc/jit/python/python_sugared_value.h (hunk 8 above)
Comment (voznesenskym) [reply torch/csrc/jit/python/python_sugared_value.h:179]: Agreed. Any suggestions?

Comment (suo) [reply torch/csrc/jit/python/python_sugared_value.h:179]: SugaredDict?

File: aten/src/ATen/core/type.cpp (hunk 4 above)
Comment (suo) [reply aten/src/ATen/core/type.cpp:1163]: curious what happens in eager if someone registers the same name/value pair as a parameter and buffer…

File: test/test_jit.py (hunk 7 above)
Comment (voznesenskym) [reply test/test_jit.py:17172]: Now it does :)

File: aten/src/ATen/core/type.cpp (hunk 4 above)
Comment (voznesenskym) [reply aten/src/ATen/core/type.cpp]: I'll add a test for this!

File: aten/src/ATen/core/jit_type.h [hunk 12]
```diff
@@ -1600,6 +1600,39 @@ CAFFE2_API TypePtr tryEvalTypeVariables(TypePtr type, TypeEnv& type_env);
...
+
+  AttributeKind getKind() const {
+    return kind_;
+  }
+
+  TypePtr getType() const {
+    return attributeType_;
+  }
+
+  const std::string getName() const {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1626]: nit: ref

File: aten/src/ATen/core/jit_type.h [hunk 13]
```diff
@@ -1600,6 +1600,39 @@ CAFFE2_API TypePtr tryEvalTypeVariables(TypePtr type, TypeEnv& type_env);
 
 CAFFE2_API bool elementTypeCanBeInferredFromMembers(const TypePtr& elem_type);
 
+enum class AttributeKind {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h:1603]: nit: this enum and class are relatively self-explantory, but a small comment may be useful for the future

File: torch/csrc/jit/api/module.h [hunk 14]
```diff
@@ -117,25 +117,31 @@ struct TORCH_API Module : public Object {
   // register_buffer method. With this simplification, we only need to track
   // whether a slot is a parameter to be able to classify it.
   void register_buffer(const std::string& name, at::Tensor v) {
-    type()->addOrCheckAttribute(name, TensorType::get());
+    bool is_param = false;
+    bool is_buffer = true;
+    type()->addOrCheckAttribute(name, TensorType::get(), is_param, is_buffer);
```
Comment (suo) [inline torch/csrc/jit/api/module.h:122]: Hmm…having a lot of boolean defaults in a row is getting a little scary. For example, there is a bug here, because you are actually setting `is_param = false, allow_any = true, is_buffer=false`, which is not what you want. I think either we should undefault them or change it to something more declarative than several flags

File: torch/csrc/jit/serialization/python_print.cpp [hunk 15]
```diff
@@ -1229,19 +1229,29 @@ struct PythonPrintImpl {
       // attributes and parameters.
       if (is_module) {
         std::vector<std::string> params;
+        std::vector<std::string> buffers;
```
Comment (suo) [inline torch/csrc/jit/serialization/python_print.cpp:1232]: Nice! Before committing this, we'll have to test for forward compatibility. Basically, can you `save()` with buffers, then `load()` using a version of pytorch that does not understand buffer serialization.

Based on skimming the import code, I think there will be an FC issue. Let's chat directly about the steps to take here.

**Review by suo — CHANGES_REQUESTED** [review]
nice, very close!

File: torch/csrc/jit/api/module.h (hunk 14 above)
Comment (voznesenskym) [reply torch/csrc/jit/api/module.h:122]: Agreed. I know you said it earlier and I pushed back, but I find myself coming around to the idea of replacing it with our enum.

Comment (voznesenskym) [reply torch/csrc/jit/api/module.h:122]: I like your earlier suggestion of utilizing our enum. Since this would be a large (but relatively simple) change, are you fine with me moving the attribute adding APIs to the new enum in a subsequent diff?

File: torch/csrc/jit/serialization/python_print.cpp (hunk 15 above)
Comment (voznesenskym) [reply torch/csrc/jit/serialization/python_print.cpp:1232]: Agreed, let me figure out the best way to do so.

File: torch/csrc/jit/api/module.h (hunk 14 above)
Comment (suo) [reply torch/csrc/jit/api/module.h:122]: yeah I think that makes sense!

Comment (suo) [reply torch/csrc/jit/api/module.h:122]: This is still a bug, right?

File: torch/csrc/jit/api/module.h [hunk 16]
```diff
@@ -117,25 +117,31 @@ struct TORCH_API Module : public Object {
...
     _ivalue()->setAttr(name, std::move(v));
   }
+
   void register_parameter(
       const std::string& name,
       at::Tensor v,
       bool is_buffer) {
-    type()->addOrCheckAttribute(name, TensorType::get(), !is_buffer);
+    type()->addOrCheckAttribute(name, TensorType::get(), !is_buffer, is_buffer);
```
Comment (suo) [inline torch/csrc/jit/api/module.h:130]: Same here

File: torch/csrc/jit/api/module.h (hunk 14 above)
Comment (voznesenskym) [reply torch/csrc/jit/api/module.h:122]: Yea, I thought I pushed a fix. will update.

File: torch/csrc/jit/api/module.h (hunk 16 above)
Comment (voznesenskym) [reply torch/csrc/jit/api/module.h:130]: Yeah, will fix. Good catch.

**Review by suo — APPROVED** [review]
looks good! I recommend sanity-checking the FC changes manually if you haven't already.

**Comment by voznesenskym** [comment]
FC tests completed via:

1. Saving models locally on current commit
2. Switching branches
3. Rebuilding pytorch
4. Loading models on older versions (b9c537514cc6d1f64208f893b609129a18e4db45)
